### DISCUSSION TITLE
Is there a way to set a custom typescript-type for a field?

### DISCUSSION THREAD

**Discussion by joshstrange** [prisma/prisma#11196] [Q&A] [answered]
I'm migrating a codebase over to use Prisma (previously all raw queries through aurora's data-api). I prefer to manage my enum-type columns in code instead of the DB so I have a varchar column for "role" and it's just a string of one of a number of roles. In my codebase I have my roles defined as such:

```
export const Role = {
	CUSTOMER: 'customer',
	ADMIN: 'admin',
	SYSTEM: 'system',
	MANAGER: 'manager',
        .....
} as const;
type TRole = keyof typeof Role;
export type TUserRole = typeof Role[TRole];
```

Previously I just typed the `role` field to be of type `TUserRole` (`getUser(...): Promise<{id: string, role: TUserRole, ..... }>`) but Prisma just returns it as a string (no surprise, that's what the column is). I'm wondering if there is anyway I can override the Typescript definition that is generated so that I can force `role` to be of type `TUserRole`. Without this I would have to cast `role` to `TUserRole` everywhere I use it or I get TS errors "TS2345: Argument of type 'string' is not assignable to parameter of type 'TUserRole'.".

Thank you for any help, I'm loving Prisma so far!

**Comment by austincrim** [comment, marked as answer]
Hey @joshstrange 👋 ,

This is not possible with Prisma today, but we do have an existing issue that would cover this use case (#7400). It would be great if you could give that issue a +1 and describe your unique situation. This helps our product and engineering teams prioritize our roadmap.

Would you mind telling me why you want to handle the enum in code instead of adding it to your `schema.prisma`? For the record, here is what it would look like

in the schema
```prisma
enum Role {
  Customer
  Admin
  System
  Manager
}
```

and in the generated client code.
```ts
export const Role: {
  Customer: 'Customer',
  Admin: 'Admin',
  System: 'System',
  Manager: 'Manager'
};

export type Role = (typeof Role)[keyof typeof Role]
```

Let me know if you have any other questions!

**Reply by joshstrange** [reply]
My main aversion to using DB enums is because I can add an "enum" value in my code that can be updated without any DB migrations needed. Prisma migrate makes life a little easier but no migration is always better. I understand my use case is outside of what you all support and that's fine. I've just wrapped/converted the `User` object I get back from prisma into my internal interface I was using before prisma. I might try DB enums in the future once I have migrations working smoothly. Thank you!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
